### PR TITLE
Lab05 - Stringers

### LINKED ISSUE #368: Lab05 - Stringers

### PR DESCRIPTION
**Pull request by willshen8**
Fixes #368 

Review of colleague's PR #363 

## - Stringers lab
- Make type IPAddr
- Implement fmt.stringer to be able to print it as a string in IP address format
- Test cases

### CODE REVIEW

File: 05_stringer/willshen8/stringer.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+var out io.Writer = os.Stdout
+
+type IPAddr [4]byte
+
+func (ip IPAddr) String() string {
+	var returnIPAddress string
+	for _, v := range ip {
+		returnIPAddress += strconv.Itoa(int(v)) + "."
+	}
+	returnIPAddress = strings.TrimSuffix(returnIPAddress, ".")
```
Comment (joel00wood) [inline 05_stringer/willshen8/stringer.go]: Given that the underlying slice is of a fixed length and of a predefined order the loop and trim could be simplified down to:
```
returnIpAddress = fmt.Sprintf("%v.%v.%v.%v", ip[0], ip[1], ip[2], ip[3])
```

**Review by joel00wood — COMMENTED** [review]
Only one minor comment around the approach to string construction, otherwise appears fit for purpose

File: 05_stringer/willshen8/stringer_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		{name: "Empty IP Address", input: IPAddr{0, 0, 0, 0}, expected: "0.0.0.0"},
+		{name: "Broadcast IP Address", input: IPAddr{255, 255, 255, 255}, expected: "255.255.255.255"},
+	}
+
+	for _, test := range tests {
+		expected := test.expected
+		input := test.input
+		t.Run(test.name, func(t *testing.T) {
+			actual := input.String()
+			assert.Equal(t, expected, actual, "Test Failed! Actual result is not equal to expected")
```
Comment (joshcarp) [inline 05_stringer/willshen8/stringer_test.go]: This doesn't add anything; failing tests will indicated that they failed

File: 05_stringer/willshen8/stringer_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	out = &buf
+	main()
+	expected := "127.0.0.1\n"
+	actual := buf.String()
+
+	assert.Equal(t, expected, actual, "Test Failed in main! Actual is not equal to expected result")
+}
+
+func TestStringer(t *testing.T) {
+	tests := []struct {
```
Comment (joshcarp) [inline 05_stringer/willshen8/stringer_test.go]: I prefer test variables to be global vars

File: 05_stringer/willshen8/stringer_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		name     string
+		input    IPAddr
+		expected string
+	}{
+		{name: "Empty IP Address", input: IPAddr{0, 0, 0, 0}, expected: "0.0.0.0"},
+		{name: "Broadcast IP Address", input: IPAddr{255, 255, 255, 255}, expected: "255.255.255.255"},
+	}
+
+	for _, test := range tests {
+		expected := test.expected
```
Comment (joshcarp) [inline 05_stringer/willshen8/stringer_test.go]: get rid of ```expected := test.expected```, use ```test.expected``` directly

File: 05_stringer/willshen8/stringer_test.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+	main()
+	expected := "127.0.0.1\n"
+	actual := buf.String()
+
+	assert.Equal(t, expected, actual, "Test Failed in main! Actual is not equal to expected result")
```
Comment (joshcarp) [inline 05_stringer/willshen8/stringer_test.go]: this one is not needed either

**Review by joshcarp — CHANGES_REQUESTED** [review]

File: 05_stringer/willshen8/stringer.go (hunk 1 above)
Comment (willshen8) [reply 05_stringer/willshen8/stringer.go]: Thanks, this is a neater approach.

File: 05_stringer/willshen8/stringer_test.go (hunk 4 above)
Comment (willshen8) [reply 05_stringer/willshen8/stringer_test.go]: Moving tests as global var seems to resolve my problem here, I had a lint warning, hence I had expected := test.expected

File: 05_stringer/willshen8/stringer_test.go (hunk 3 above)
Comment (willshen8) [reply 05_stringer/willshen8/stringer_test.go]: I agree, as these can be reused if necessary.

File: 05_stringer/willshen8/stringer_test.go (hunk 5 above)
Comment (willshen8) [reply 05_stringer/willshen8/stringer_test.go]: Thanks, resolved.

File: 05_stringer/willshen8/stringer_test.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+	assert.Equal(t, expected, actual)
+}
+
+var tests = []struct {
+	name     string
+	input    IPAddr
+	expected string
+}{
+	{name: "Empty IP Address", input: IPAddr{0, 0, 0, 0}, expected: "0.0.0.0"},
```
Comment (joshcarp) [inline 05_stringer/willshen8/stringer_test.go:26]: Probably good to put more testcases in ```IPAddr{127,0,1}```

**Review by joshcarp — CHANGES_REQUESTED** [review]

**Review by joshcarp — APPROVED** [review]

**Pull request merged**